src/edu/washington/escience/myria/operator/Sample.java

```diff
@@ -0,0 +1,241 @@
+
+  /** Number of tuples to sample from the right operator. */
+  private int sampleSize;
+
+  /** Random generator used for index selection. */
+  private Random rand;
+
+  /** The type of sampling to perform. Currently supports 'WR' and 'WoR'. */
+  private String sampleType;
+
+  /** True if operator has extracted sampling info. */
+  private boolean computedSamplingInfo = false;
+
+  /** Buffer for tuples that will be returned. */
+  private TupleBatchBuffer ans;
+
+  /** Global count of the tuples seen so far. */
+  private int tupleNum = 0;
+
+  /** Sorted array of tuple indices that will be taken as samples. */
+  private int[] samples;
+
+  /** Current index of the samples array. */
+  private int curSampIdx = 0;
+
+  /**
+   * Instantiate a Sample operator using sampling info from the left operator
+   * and the stream from the right operator.
+   *
+   * @param left
+   *          inputs a (WorkerID, StreamSize, SampleSize, SampleType) tuple.
+   * @param right
+   *          tuples that will be sampled from.
+   * @param randomSeed
+   *          value to seed the random generator with. null if no specified seed
+   */
+  public Sample(final Operator left, final Operator right, Long randomSeed) {
+    super(left, right);
+    this.rand = new Random();
+    if (randomSeed != null) {
+      this.rand.setSeed(randomSeed);
+    }
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws Exception {
+    // Extract sampling info from left operator.
+    if (!computedSamplingInfo) {
+      TupleBatch tb = getLeft().nextReady();
+      if (tb == null)
+        return null;
+      extractSamplingInfo(tb);
+      getLeft().close();
+
+      // Cannot sampleWoR more tuples than there are.
+      if (sampleType.equals("WoR")) {
+        Preconditions.checkState(sampleSize <= streamSize,
+            "Cannot SampleWoR %s tuples from a population of size %s",
+            sampleSize, streamSize);
+      }
+
+      // Generate target indices to accept as samples.
+      if (sampleType.equals("WR")) {
+        samples = generateIndicesWR(streamSize, sampleSize);
+      } else if (sampleType.equals("WoR")) {
+        samples = generateIndicesWoR(streamSize, sampleSize);
+      } else {
+        throw new DbException("Invalid sampleType: " + sampleType);
+      }
+
+      computedSamplingInfo = true;
+    }
+
+    // Return a ready tuple batch if possible.
+    TupleBatch nexttb = ans.popAny();
+    if (nexttb != null) {
+      return nexttb;
+    }
+    // Check if there's nothing left to sample.
+    if (curSampIdx >= samples.length) {
+      getRight().close();
+      setEOS();
+      return null;
+    }
+    Operator right = getRight();
+    for (TupleBatch tb = right.nextReady(); tb != null; tb = right.nextReady()) {
+      if (curSampIdx >= samples.length) { // done sampling
+        break;
+      }
+      if (samples[curSampIdx] > tupleNum + tb.numTuples()) {
+        // nextIndex is not in this batch. Continue with next batch.
```

Comment:
Isn't this short-circuit unnecessary since the while-loop condition will be initially unsatisfied and we increment `tupleNum` and hit the bottom of the loop? The only difference is we don't check for a filled `TupleBatch` in this codepath, which shouldn't matter since we didn't add any results in this iteration.